flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -1149,6 +1164,53 @@ mlir::Value IntrinsicLibrary::genAimag(mlir::Type resultType,
       args[0], true /* isImagPart */);
 }
 
+// ALL
+fir::ExtendedValue
+IntrinsicLibrary::genAll(mlir::Type resultType,
+                         llvm::ArrayRef<fir::ExtendedValue> args) {
+
+  assert(args.size() == 2);
+  // Handle required mask argument
+  auto mask = builder.createBox(loc, args[0]);
+
+  fir::BoxValue maskArry = builder.createBox(loc, args[0]);
+  int rank = maskArry.rank();
+  assert(rank >= 1);
+
+  // Handle optional dim argument
+  bool absentDim = isAbsent(args[1]);
+  auto dim = absentDim
+             ? builder.createIntegerConstant(loc, builder.getIndexType(), 1)
+             : fir::getBase(args[1]);
+
+   if (rank == 1 || absentDim)
+     return builder.createConvert(
+        loc, resultType,
+        Fortran::lower::genAll(builder, loc, mask, dim));
+
+  // else use the result descriptor AllDim() intrinsic
+
+  // Create mutable fir.box to be passed to the runtime for the result.
+
+  auto resultArrayType = builder.getVarLenSeqTy(resultType, rank - 1);
+  auto resultMutableBox =
+       Fortran::lower::createTempMutableBox(builder, loc, resultArrayType);
+  auto resultIrBox =
+       Fortran::lower::getMutableIRBox(builder, loc, resultMutableBox);
+
+  // Call runtime. The runtime is allocating the result.
+  Fortran::lower::genAllDescriptor(builder, loc, resultIrBox, mask, dim);
+  return Fortran::lower::genMutableBoxRead(builder, loc, resultMutableBox)
+      .match(
+          [&](const fir::ArrayBoxValue &box) -> fir::ExtendedValue {
+            return box;
```

Comment:
The clean-up (that insert deallocation (and finalization in the future), after the statement) is still needed here, it looks like you removed it (`addCleanUpForTemp`).